### ISSUE TITLE
bias gradient goes to inf

### ISSUE THREAD

**Issue by yhcc** [HazyResearch/flash-attention#117]
I don't know why when i tried to run GPTModel, the gradient of its bias term will go to infinity.

```python
#!/usr/bin/env python
# -*- encoding: utf-8 -*-
# torchrun --standalone --nproc_per_node 4 --nnodes 1 torch_debug.py
from functools import partial
import torch
import torch.multiprocessing as mp
import numpy as np

from torch import nn
from flash_attn.losses.cross_entropy import CrossEntropyLoss as FlashCrossEntropyLoss
from flash_attn.models.gpt import GPTLMHeadModel
from transformers.models.gpt2.configuration_gpt2 import GPT2Config


class FlashGPTLMLoss(nn.Module):

    def __init__(self, pg, parallel_output=True):
        super().__init__()
        if parallel_output:
            self.loss_fn = FlashCrossEntropyLoss(reduction='mean', inplace_backward=True, process_group=pg)
        else:
            self.loss_fn = nn.CrossEntropyLoss(reduction='mean') 

    def forward(self, logits, labels, **kwargs):
        shift_logits = logits[..., :-1, :].contiguous().view(-1, logits.size(-1))
        shift_labels = labels[..., 1:].contiguous().view(-1)
        loss = self.loss_fn(shift_logits, shift_labels) 

        return loss


def check_layer():
    from colossalai.initialize import launch
    from colossalai.core import global_context as gpc
    from colossalai.context.parallel_mode import ParallelMode
    import os

    CONFIG = dict(parallel=dict(pipeline=dict(size=1), tensor=dict(size=4, mode='1d')))

    # I just use this colossalai to launch tensor parallelism progress group
    launch(config=CONFIG, rank=int(os.environ['LOCAL_RANK']), world_size=int(os.environ['WORLD_SIZE']), host='localhost', port=7777, backend='nccl')
    rank = torch.distributed.get_rank()
    print("Rank is ", rank)
    pg = gpc.get_group(ParallelMode.TENSOR)
    
    rank = torch.distributed.distributed_c10d._get_global_rank(pg, rank)
    print("Rank is ", rank)
    pg = torch.distributed.distributed_c10d._get_default_group()
    seqlen = 2048
    hidden_dim = 12288//2
    nheads = 96//2
    n_layer = 4
    rotary_emb_fraction = 0.5
    config = GPT2Config(vocab_size=50257, n_positions=seqlen, n_embd=hidden_dim,
                        n_layer=n_layer, n_head=nheads, 
                        scale_attn_by_inverse_layer_idx=True, 
                        rotary_emb_fraction=rotary_emb_fraction,
                        use_flash_attn=True, fused_mlp=True,
                        fused_bias_fc=True, fused_dropout_add_ln=True, 
                        pad_vocab_size_multiple=8)
    model = GPTLMHeadModel(config, process_group=pg, device=torch.device('cuda'), dtype=torch.half).half().cuda()
    rng = np.random.RandomState(0)  # make sure all rank with the same input
    input_ids = torch.from_numpy(rng.randint(0, 100, size=(1, 2048))).cuda()
    output = model(input_ids=input_ids).logits
    loss = FlashGPTLMLoss(pg=pg)(output, input_ids)

    print('loss', loss)
    total_inf = 0
    loss.backward()
    for name, param in model.named_parameters():
        total_inf += param.grad.isinf().sum().item()
        if param.grad.isinf().sum().item()>0:
            print(rank, name, param.grad.isinf().sum().item())

    print(f"In rank:{torch.distributed.get_rank()}, found {total_inf} infs")


if __name__ == '__main__':
    check_layer()
```
Some demo output will be like this
```
tensor(11.8527, device='cuda:3', grad_fn=<DivBackward0>)

tensor(11.8527, device='cuda:2', grad_fn=<DivBackward0>)
0 transformer.layers.2.mixer.out_proj.bias 1
In rank:1, found 0 infs
In rank:2, found 0 infs
In rank:3, found 0 infs
In rank:0, found 1 infs
```
The loss is ok, but inf detacted in bias term

**Comment by yhcc** [comment]
It turns out this seems to be caused by the cuda version mismatch between torch and flashattention

**Comment by zhuzn** [comment]
Please how do you install xentropy_cuda_lib when you use FlashCrossEntropyLoss?

**Comment by tridao** [comment]
> Please how do you install xentropy_cuda_lib when you use FlashCrossEntropyLoss?

Responded in #128
